### ISSUE TITLE
FoldChange results depend on the group size

### ISSUE THREAD

**Issue by roi-meir** [satijalab/seurat#9346]
Hi

Seurat recently changed the pseudocount used in the log fold change computation. Instead of using a fixed pseudocount, it is divided by the group size. This can cause a a different log fold change that depends on the group's size even if the mean is the same.
The log fold change results will prefer the smaller group because it's pseudocount will be smaller 

The code appears in multiple locations but for example from [differential_expression.R](https://github.com/satijalab/seurat/blob/master/R/differential_expression.R#L1073)
```r
counts.mean.fxn <- function(x) {
    # return(log(x = rowMeans(x = x) + pseudocount.use, base = base))
    return(log(x = (rowSums(x = x) + pseudocount.use)/NCOL(x), base = base))
  }
```

I think that we should add a small constant epsilon as a pseudocount (as done in Scanpy)

To reproduce it I created a small dataset with equal expression in all cells and got log fold change depending on the group size. 

```r
library(dplyr)
library(Seurat)

obj <- CreateSeuratObject(data.frame(x=rep(0.1, 10), y=rep(1, 10), z=rep(10, 10), row.names = 1:10) %>% t)
Idents(obj) <- colnames(obj)
message("5 vs 2")
FoldChange(obj, ident.1 = 1:5, ident.2 = 6:7, slot = "counts")
message("5 vs 4")
FoldChange(obj, ident.1 = 1:5, ident.2 = 6:9, slot = "counts")
message("5 vs 5")
FoldChange(obj, ident.1 = 1:5, ident.2 = 6:10, slot = "counts")
```

Output 
```
5 vs 2
   avg_log2FC pct.1 pct.2
x -1.00000000     1     1
y -0.32192809     1     1
z -0.04182018     1     1
5 vs 4
    avg_log2FC pct.1 pct.2
x -0.222392421     1     1
y -0.058893689     1     1
z -0.007054758     1     1
5 vs 5
  avg_log2FC pct.1 pct.2
x          0     1     1
y          0     1     1
z          0     1     1
```

sessionInfo:
```
R version 4.3.1 (2023-06-16)
Platform: x86_64-conda-linux-gnu (64-bit)
Running under: Gentoo/Linux

Matrix products: default
BLAS/LAPACK: ~/miniconda/envs/test_env/lib/libopenblasp-r0.3.21.so;  LAPACK version 3.9.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C
 [9] LC_ADDRESS=C               LC_TELEPHONE=C
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base

other attached packages:
[1] Seurat_5.1.0       SeuratObject_5.0.2 sp_2.1-4           dplyr_1.1.3

loaded via a namespace (and not attached):
  [1] deldir_2.0-4           pbapply_1.7-2          gridExtra_2.3
  [4] rlang_1.1.1            magrittr_2.0.3         RcppAnnoy_0.0.22
  [7] spatstat.geom_3.3-3    matrixStats_1.4.1      ggridges_0.5.6
 [10] compiler_4.3.1         png_0.1-8              vctrs_0.6.4
 [13] reshape2_1.4.4         stringr_1.5.0          pkgconfig_2.0.3
 [16] fastmap_1.1.1          ellipsis_0.3.2         utf8_1.2.4
 [19] promises_1.2.1         purrr_1.0.2            jsonlite_1.8.7
 [22] goftest_1.2-3          later_1.3.1            spatstat.utils_3.1-0
 [25] irlba_2.3.5.1          parallel_4.3.1         cluster_2.1.4
 [28] R6_2.5.1               ica_1.0-3              stringi_1.7.12
 [31] RColorBrewer_1.1-3     spatstat.data_3.1-2    reticulate_1.39.0
 [34] parallelly_1.36.0      spatstat.univar_3.0-1  lmtest_0.9-40
 [37] scattermore_1.2        Rcpp_1.0.11            tensor_1.5
 [40] future.apply_1.11.0    zoo_1.8-12             sctransform_0.4.1
 [43] httpuv_1.6.11          Matrix_1.6-4           splines_4.3.1
 [46] igraph_2.0.3           tidyselect_1.2.0       abind_1.4-8
 [49] spatstat.random_3.3-2  codetools_0.2-19       miniUI_0.1.1.1
 [52] spatstat.explore_3.3-2 listenv_0.9.0          lattice_0.22-5
 [55] tibble_3.2.1           plyr_1.8.9             shiny_1.7.5.1
 [58] ROCR_1.0-11            Rtsne_0.17             future_1.33.0
 [61] fastDummies_1.7.4      survival_3.5-7         polyclip_1.10-7
 [64] fitdistrplus_1.2-1     pillar_1.9.0           KernSmooth_2.23-22
 [67] plotly_4.10.4          generics_0.1.3         RcppHNSW_0.6.0
 [70] ggplot2_3.4.4          munsell_0.5.0          scales_1.2.1
 [73] globals_0.16.2         xtable_1.8-4           glue_1.6.2
 [76] lazyeval_0.2.2         tools_4.3.1            data.table_1.14.8
 [79] RSpectra_0.16-2        RANN_2.6.2             leiden_0.4.3.1
 [82] dotCall64_1.1-1        cowplot_1.1.3          grid_4.3.1
 [85] tidyr_1.3.0            colorspace_2.1-0       nlme_3.1-163
 [88] patchwork_1.3.0        cli_3.6.1              spatstat.sparse_3.1-0
 [91] spam_2.10-0            fansi_1.0.5            viridisLite_0.4.2
 [94] uwot_0.2.2             gtable_0.3.4           digest_0.6.33
 [97] progressr_0.14.0       ggrepel_0.9.6          htmlwidgets_1.6.4
[100] farver_2.1.1           htmltools_0.5.8.1      lifecycle_1.0.3
[103] httr_1.4.7             mime_0.12              MASS_7.3-60
```

**Comment by roi-meir** [comment]
I've opened a pull request to fix this issue

**Comment by cyrillustan** [comment]
Copied my comment to your PR here:
I can agree with your change of taking the average first and then taking the logarithm, but why do you want to change the pseudocount to 1e-06? expm1() and log1p() are inverse functions. On a linear scale, having a pseudocount of 1e-06 and 1e-10 does not make a difference, but on a logscale it actually changes the values greatly.

**Comment by roi-meir** [comment]
As mentioned in other issues it may cause a group with lower expression to show a higher log fold change   
For example here the first group expression is all 0.06 while the second is all 0.04 but still, we get a negative log fold change 
```r

library(dplyr)
library(Seurat)

obj <- CreateSeuratObject(data.frame(x=c(rep(0.06, 6), rep(0.04, 4)), y=rep(1,10), row.names = 1:10) %>% t)
Idents(obj) <- colnames(obj)
FoldChange(obj, ident.1 = 1:6, ident.2 = 7:10, slot = "counts")
# Output
   avg_log2FC pct.1 pct.2
x -0.35548065     1     1
y -0.09953567     1     1
```

**Comment by cyrillustan** [comment]
Thank you for your thoughtfulness! After some discussion in the group, we decided not to change this for now.
In the recent change we made, we elected to add the pseudocount on a group level. With this choice, group size will influence the results slightly unfortunately, which leads to changes in fold changes too. However, fold change is intrinsically an unstable value for lowly expressed genes, and one has to look at the results in the context of the p-value. In these cases, we suggest not focusing too much on the fold change values themselves but on the underlying meaning.
